### ISSUE TITLE
Consider dropping `non_exhaustive` on `Network`, etc

### ISSUE THREAD

**Issue by TheBlueMatt** [rust-bitcoin/rust-bitcoin#2225]
In 0.30, `#[non_exhaustive]` was added to the `Network` enum in what I assume was a step towards stabilization with forwards compatibility. `#[non_exhaustive]` doesn't, itself, provide forwards compatibility with minor version changes, but rather forces downstream devs to think about it, and puts the burden for forwards compatibility on them. This only works if downstream projects have to be able to have some kind of sensible default behavior for "unknown future bitcoin-style network", which isn't super practical. For example, this came up in the `lightning-invoice` crate, we have to convert from the rust-bitcoin `Network` to a BOLT11-specific `Currency` - we cannot reasonably have forward-compatible code that will "just work" if a new enum variant is added, so an [`unreachable` was added](https://github.com/lightningdevkit/rust-lightning/blob/ad56847/lightning-invoice/src/lib.rs#L410). That will obviously be removed, but there's really nothing we can do to be forwards compatible here, and instead would strongly prefer to see a version bump of rust-bitcoin so that we can update our code to handle entire new bitcoin-compatible networks.

**Comment by apoelstra** [comment]
Specifically with `Network` we've moved away from an open-ended enum to just a "mainnet/testnet" dichotomy. So definitely +1 to removing `#[non_exhaustive]` there.

For other types, I tend to agree with you. I think we should leave `#[non_exhaustive]` on errors (which we expect people will rarely/never match on, and usually only match on a single variant that they care about), and for everything else we should think hard about what we actually expect people to do with unknown variants. And if the answer is "panic" we shouldn't make them do that.

**Comment by clarkmoody** [comment]
Agreed on removal where it makes sense. I'm not a huge fan of `non_exhaustive` generally, but it does make sense for parts of a library API.

**Comment by Kixunil** [comment]
NACK, I believe having the enum `#[non_exhaustive]` makes more sense. As we can already observe, there's wide variety of applications supporting different networks. Electrum supports `simnet` (IDK what it is) in addition to the networks we support for instance, some applications do not support the existing networks and others are only defined for the current ones and it's unclear what will happen when there's a new one. For instance currency in LN is defined for each network we have today except for simnet. It's not predetermined what will happen once this changes.

As such I believe the parameter is application-specific and cannot be universal unless the application is specifically built to support all possible networks. (This is possible, I did this for Firefish by using magic as the network identifier when serialized and native parsing methods for stringly inputs.) And if an application is not made universal then it has to have its own type to represent the network **it** supports, with conversions and such.

I'm aware this is more annoying but like many other things in Rust, it prevents problems by dealing with these things upfront. I'm still interested in making the API easier. For instance I think using `impl Into<Network>` in function signatures may be a good idea, just must be weighted against loss of inference.

Another approach we could take for selected applications is to add methods that convert values to and from their representation. E.g. we could add methods to return LN currency string. **But** the conditions for inclusion would be very strict: we would have to be convinced that whenever Core releases support for a new network the appropriate teams will define how this is mapped to their domain ASAP. So e.g. if Core adds `foonet` then LN folks will quickly decide what the string for currency is. FTR I currently don't know of any case where I would be confident to do that.

We could also provide a macro that makes it easier to define your own network types if it helps.

**Comment by Kixunil** [comment]
They already suffer the issue, they just wouldn't need to convert between networks inside business rules but at the boundary which is good. I don't see how it's related to strings, since the crates can (and should) define their own strong types.

Also if there's a collection of crates that is guaranteed to support some specific set of networks they should share a type.

**Comment by Kixunil** [comment]
> it forces users to deal with an _unknown network_ up front

It doesn't *force* users to do that. It's entirely possible to design an application that uses no `match` and communicates with external entities using already-standard types/interfaces (magic, human-readable names). I know this because I've done it in an actual production application. Also IIRC elects is written this way too.

The fact that you perceive it as forcing stems from you happening to work on a library where the protocol itself forces you to do something different. And because you're forced to use a specific set of networks you should define your own - which you did with the `Currency` type. If you impl `From<Currency> for Network` then you nearly won't see a big difference. Calling `into()` sometimes is fine, we can remove it completely trading it for worse inference.

For the opposite conversion, I have verified that we never return `Network` from any function. (LMK if I missed something.) It's always `Option<Network>` or `Result<Network>` this means that for any case where you'd be forced to deal with an unknown network you would be regardless! You can just map whatever you don't support to `None`/`Err`:

```rust
let currency = match Network::from_chain_hash(chain_hash) {
    Some(Network::BItcoin) => Currency::Bitcoin,
    Some(Network::Regtest) => Currency::Regtest,
    Some(Network::Testnet) => Currency::Testnet,
    Some(Network::Signet) => Currency::Signet,
    _ => return Err(UnknownChainHash(chain_hash)),
}
```

The code is pretty elegant IMO.

> If we remove `#[non_exhaustive]` that's not a case that ever exists, a change to rust-bitcoin to add a new network will simply require handling a new network

It's an unknown network for the protocol that doesn't define it. But what's worse it will force breakage across the entire ecosystem and cause coordination nightmares whenever a new network is added. We're trying to bring the crate to the state where each crate can rely on 1.0 primitives and never worry about breakages again. Isn't it worth a few more lines in your code?

> Another way to look at rust-bitcoin's `Network` is "the minimum set of supported networks that everyone should really support",

I disagree with this view in two ways. First, I personally see very little value in supporting anything else than mainnet and regtest for most applications. Regtest is million times easier to work with than everything else when it comes to testing. So I don't understand why the type should have other variants if this is its definition. Second, the type really is "whatever Core supports" this can be clearly seen from having `to_core_arg` (infallible) and `from_core_arg` which suggests that we've *committed* to never adding a network that Core doesn't support. The other conversions are also defined in Core (but it's consensus, so it doesn't matter much).

**Comment by Kixunil** [comment]
> Sadly Bitcoin is filled with people defining their own network enum.

Really? Can you point to specific examples? I haven't seen any other than `Currency` in lightning which is perfectly reasonable.

> not having good convertibility between them would be a massive PITA for everyone (I think on this point we agree)

It already is by simply the applications and, more importantly, protocols not being designed to be network-agnostic. You can't change it by making the enum not `#[non_exhaustive]` because then we'll need to issue breaking releases and you can't convert between different versions of `Network` so we're back at square one and even worse - it infects all other types in the crate so the whole crate becomes incompatible!

> Users are going to use the stuff that's exposed, and not always in the idealized way

I'm willing to accept it for many other things. But after fighting so hard to get this crate to stabilize one day being told that's not going to happen is absolutely horrible to hear. If that's the position of other maintainers I might be better off forking the crate.

> given the BOLT 11 set is congruous with the rust-bitcoin set

Only today. It's perfectly possible it won't be in the future. E.g. I have an example of a network that I believe could be interesting: signed like signet but zero difficulty like regtest. And when that happens we get ecosystem-wide breakage forcing everyone to update or be incompatible.

> We store a `Network` in some `Invoice`

Yeah, that's the incorrect thing. You should store `Currency`. Basically use `Currency` everywhere except when communicating with the `bitcoin` crate. Then it works fine. This is the classic Rust pattern that you validate data at boundary and use strong type inside so I'm not even proposing anything exotic.

**Comment by tcharding** [comment]
Can this debate be reduced to the following question:

- `rust-bitcoin v1.0.0` is released.
- Bitcoin Core adds a new network.
- Should users of `rust-bitcoin` be able to use `v1.1.0` to get the new network or is it reasonable to force them to use `v2.0.0` to get it (forcing them to take all the other changes in `v2.0`)?

Its a trade off, right? Upgrade to `v2.0.0` might be super painful (or timewise impossible) but enabling usage of `v1.1.0` is painful because it requires `non_exhaustive` on the `Network` type.

**Comment by Kixunil** [comment]
@TheBlueMatt I took a look at LDK and I found that nothing is broken, just `From` needed to be changed to `TryFrom` so it looks like it's not as big of a deal as you thought? Did I miss something? See this PR: https://github.com/lightningdevkit/rust-lightning/pull/2789

I've searched for other instances of matching on `Network` and didn't find any. So it seems your crate is completely unaffected aside from `Currency` being a separate type which may be annoying to some people.

@tcharding yes, that summary is accurate, though maybe `#[non_ehaustive]` is not *that* painful. I've yet to see a crate where it causes a massive breakage.

**Comment by Kixunil** [comment]
> but is also likely to impact other things, like how work calculations work across blocks, what difficulty means, etc

Those are entirely contained within `Params` and our API methods and will be even more when we implement proper difficulty adjustment. Moreover, many, many applications don't really need to deal with these things as they simply connect to Core and accept it as the source of truth.

> Great, so now we've pushed the dead-code error handling from rust-bitcoin down into LDK and then pushed it further into the code of our users, rather than just not having to handle the non-case at all :)

They already have to handle invalid inputs when they parse and if they use `Network` instead they only have to change it to `Currency`.

> If we think this is a reasonable use of the `Network` enum then istm a new variant _should_ be a major release, irrespective of semver concerns, because it implies downstream libraries (though not really applications) need to handle the new network.

I don't think so. Either the library is written to work with any network in which case they can be updated with no code change at all or the library for some reason cannot support other networks in which case it will just return errors for the unknown ones and the applications using them will propagate those errors. Nothing really has to be broken.

I was thinking of putting the network into a separate crate (with whatever other things are likely to break for stupid reasons like someone else releasing something) and then make the crate optional everywhere but I don't think it works since it forces `Address` to release a new major version anyway which still breaks things.

Anyway, since you're annoyed about `Currency` and it's only related to parsing or serializing BOLT11 invoices maybe it'd help you if we just added the methods to the `Network`. However to accept it I'd want to see several LN protocol developers to agree that in the event of a new network added in Core they will either define what the new values are within one month of Core release or leave it up to rust-bitcoin developers to decide if they can't in time. :)

**Comment by apoelstra** [comment]
No library can be written to handle "any hypothetical future network" but it's easy in most cases to write a library that supports all the existing networks. I agree with Matt that adding new networks is (a) extremely rare, and (b) likely to break things beyond the enum, even if those are things that 99% of users will never touch.

So by making the enum `#[non_exhaustive]` and forcing people to use `TryFrom` for conversions that are actually infallible, we're just creating dead error paths that are untestable until an undefined future date.

**Comment by Kixunil** [comment]
> No library can be written to handle "any hypothetical future network"

Funny that you say this because I wrote such library. :) It is perfectly possible if there's no silly protocol that would invent its own representations for networks. In my case I encode `Network` by converting it to `Magic` and back and on the side of WASM consumers I just take and parse strings because JavaScript can't reasonably use anything else anyway.

I think we can decouple `Network` at least from `units`  and blockdata (or just transaction) If we do that then I'm more open to it since that would minimize the damage. That being said, `Address` is also a significant API piece that happens to use `Network` - theoretically, we're now moving away from it and since we're using `impl Trait` in arguments we might be able to separate that one as well.

Still it sucks that any software that did this right will get punished because majority of stuff is badly designed.

**Comment by apoelstra** [comment]
> It is perfectly possible if there's no silly protocol that would invent its own representations for networks. In my case I encode `Network` by converting it to `Magic` and back and on the side of WASM consumers I just take and parse strings because JavaScript can't reasonably use anything else anyway.

This sounds like you throw away the type information from `Network`, then getting it back is necessarily fallible, giving you an error path that you can use to handle unknown networks. But if somebody wasn't round-tripping to magics or strings, they could write code without error paths. Unless the `Network` enum were tagged `#[non_exhaustive]`, that is. Then they would need error paths that were impossible to execute until a probably-never-coming version of rust-bitcoin finally adds a new variant.

**Comment by Kixunil** [comment]
> then getting it back is necessarily fallible

That's only done during parsing which is already fallible so it doesn't matter (and the compiler will easily remove the dead branch).

> But if somebody wasn't round-tripping to magics or strings, they could write code without error paths.

Is there an actual real code that needs that though? Even LDK, a super-big crate, didn't have a single line that needed to be changed apart `From` -> `TryFrom`. And what are people doing with `Network` if not *eventually* translating it to something else?

**Comment by yancyribbens** [comment]
Wouldn't adding a new network be tantamount to IP creating a new version IPv4 vs IPv6?  I think most examples of parsing IP Addr type consider it to be exhaustive since the addition of a new type of IP scheme is a big deal and should cause breakage downstream.  IE not expected to happen without serious consideration.

**Comment by Kixunil** [comment]
@yancyribbens no. This adding of new networks happened in the past several times.

**Comment by tcharding** [comment]
#1930 is more-or-less the same is this issue.

**Comment by tcharding** [comment]
Is it worth reaching consensus on this? Feels like if we do nothing that is agreeing with @Kixunil. Do you guys have the energy or want to just leave it as is? Some ideas:

- We leave it on there - (what @Kixunil wants) 
- We remove it - (what @TheBlueMatt and at least one other user wants (#1930))
- We remove it and discuss putting it back on before v1.0
- @Kixunil and @TheBlueMatt punch on until one of them changes their mind.
- @apoelstra swoops in as BDFL and makes a call.

FTR I have no horse in this race.

**Comment by Kixunil** [comment]
I can change my mind if someone figures out a good solution to resolving compatibility issues when a new network is added. Basically what I want is for `Network` to be in a separate rate `bitcoin-network-enum` and when it bumps version existing stuff is mostly going to work - only the crates that depend on the new one will have to add variants when they start depending on it.

Note though that this needs to include inquiry about how much are downstream maintainers willing to *not* use `Network` in public API.

**Comment by apoelstra** [comment]
What if, to the extent possible, we used open-ended types like `network::Params` or `Hrp`. Our methods would take `T: Into<Hrp>` and our network enum would impl `Into` for all the relevant open-ended types.

Then we could make `Network` a closed enum, people could use it like normal, but whenever a new network arrives, *people using that network* need to explicitly construct a `Params` or `Hrp` or whatever to get the correct parameters. `Network` just wouldn't work for them.

We could even add a `Network2` enum which had the extra variant and again supported `Into<>` for all the open-ended types.

I recall rejecting this idea a long time ago, I think because we originally proposed using `Params` for everything which is a big heavy type, but if we instead split it up I think it'd be alright.

**Comment by Kixunil** [comment]
Yes this could work, especially if we add `Network` into `bitcoin-network-enum` crate so instead of having `Network2` we could just have the crate versioned.

The trouble is people will likely use `Network` in their APIs and then cause libcplocalypse-like event themselves. `non_exhaustive` `Network` forces them not to do that (but yeah, maybe whatever, just document it and let it be; I just wouldn't depend on such crates myself).

> using `Params` for everything which is a big heavy type

Non-issue if it's behind a reference which can also be `'static`.

**Comment by apoelstra** [comment]
> Yes this could work, especially if we add `Network` into `bitcoin-network-enum` crate so instead of having `Network2` we could just have the crate versioned.

This is morally more correct and elegant, but honestly it's less discoverable and I'd prefer doing the ugly thing where there's clearly a `Network` and `Network2` which show up beside each other in the docs (and which your favorite crate has suddenly started using) and you know the latter is the one that supports the most networks.

We can also do both -- have the crate with a versioned `Network` type and then re-export it as `Network`, `Network2`, etc. Though that's maybe the worst of both worlds .. or even worse :).

> The trouble is people will likely use `Network` in their APIs and then cause libcplocalypse-like event themselves. `non_exhaustive` `Network` forces them not to do that (but yeah, maybe whatever, just document it and let it be; I just wouldn't depend on such crates myself).

Well, it forces them to libcocalypse but at runtime, which will be worse.

But I don't think it'll actually be apocalyptic. The core types in this crate will work with every version of `Network` using the `Into<Params>` thing so addresses, xkeys, block header validation, network magics, etc etc etc will all work seamlessly. And if LDK wants to standardize on `Network` explicitly they can have a mini-apocalypse within their own set of crates which is smaller and closer to the application level.

This also lets them have their own `Network` enum which'll work with rust-bitcoin as long as they implement the appropriate `From` adaptors. Which is what I expect they'll do if we try to get too clever with our own enum ;).

> Non-issue if it's behind a reference which can also be `'static`.

Hmm, there was definitely *some* reason that we rejected this idea in the past. I seem to recall wanting a non-`'static` lifetime and then people hated the lifetime proliferation. But even if we wanted to do that (unclear; it would let people do bitcoin-inquisition stuff with custom chain params for example) we could accomplish it by using `AsRef` rather than `Into`, which hides the references away.

Anyway the previous discussion was definitely not this in-depth so I don't think it's worth the effort to dig it up.

**Comment by Kixunil** [comment]
> it's less discoverable

There's a version number and we can explicitly explain what's going on.

> maybe the worst of both worlds .. or even worse

Yeah, I suspect so. My main intention is not having to maintain multiple versions indefinitely.

> Well, it forces them to libcocalypse but at runtime, which will be worse.

It wouldn't if done correctly but if everyone understood how and why they should do it we wouldn't be having this conversation so I guess you're right.

> addresses, xkeys, block header validation, network magics, etc etc etc will all work seamlessly.

A long as we don't *return* the type. Which IIRC we're not doing already except for methods on `Network` itself.

> I seem to recall wanting a non-`'static` lifetime and then people hated the lifetime proliferation.

I think that combination of producing `&'static` but accepting any lifetime should work fine.

> it would let people do bitcoin-inquisition stuff with custom chain params for example

I don't see how this is harmful but anyway, I think `Params` really, really should be `non_exhaustive` because adding more params is likely. I've seen some things scattered around the code that should go there.

**Comment by Kixunil** [comment]
Oh, one more thing. Even if we do figure out some workaround we need to be careful about breaking the set of users that want to just `cargo update` and get a new network support out of the box. I don't know if this set is larger than one (myself) but it might be.

**Comment by apoelstra** [comment]
> Even if we do figure out some workaround we need to be careful about breaking the set of users that want to just `cargo update` and get a new network support out of the box.

If this is possible, it will be because the new network behaves substantially similar to the other networks (as far as the software in question is concerned). And in this case the `Into` generics will mean that things Just Work. Without having untestable "unknown network" codepaths which are impossible to predict the meaning of.

I think that conceptually this `Into` thing describes our differing requirements pretty well.

**Comment by Kixunil** [comment]
> because the new network behaves substantially similar to the other networks

Because it has same scripts, transaction layout etc, what most applications care about. Most applications delegate validation to Core (as they should) so the difference in behavior is similar.

> in this case the `Into` generics will mean that things Just Work

It won't because of parsing/from magic construction so if the enum is exhaustive I also have to update Cargo.toml to get the new version of `Network` (or worse, the code if we choose `Network2`). I suspect we should have both. `bitcoin-network-exhaustive` and `bitcoin-network-non-exhaustive`. I will be happy to use the former, others will use the latter.

> I don't really buy that that's realistic.

Of course it isn't, there's the time overflow bug that will have to be fixed one day... But if we want to break the API of a core primitive that causes libcpocalypse-like event there should be a pretty damn good reason to do that and "Core added a new feature we were able to anticipate" doesn't seem like one.

> we're invariably going to end up with _some_ major changes to the API in places where it least makes more sense to break compatibility and just do a major rev

That may happen but it doesn't invalidate the idea that we can minimize those cases by good API design.

**Comment by apoelstra** [comment]
> Because it has same scripts, transaction layout etc, what most applications care about. Most applications delegate validation to Core (as they should) so the difference in behavior is similar.
>

It might. Though we can't know that ahead of time.

As an aside, most applications in my experience don't do validation at all, but construction. And when constructing Bitcoin objects nobody is going to delegate to Core, they expect the library to produce valid objects.

> I suspect we should have both.

Sure, we can have both. I don't think we need extra crates for them -- the `#[non_exhaustive]` one is inherently unversioned and everybody who wants the exhaustive one expects it to come with rust-bitcoin 2.0 anyway.

I also really don't think there will be any libcopalypse event in any case. Basically no functions produce this enum, and the ones that receive it can accept a generic which describes the features they expect from a network (e.g. to have a `Hrp` for segwit addresses).

Yes, *applications that want to support the new network* will need to update their code, but that's what we want, rather than having the new network suddenly show up because of a minor version update, alongside a ton of assumptions about its behavior that were made before the network was ever imagined.

**Comment by Kixunil** [comment]
> most applications in my experience don't do validation at all, but construction.

Exactly.

> they expect the library to produce valid objects.

And we will.



> I don't think we need extra crates for them -- the `#[non_exhaustive]` one is inherently unversioned and everybody who wants the exhaustive one expects it to come with rust-bitcoin 2.0 anyway.

They need appropriate home in different crates because they are versioned differently. I personally don't plan to depend on `bitcoin` but on its individual parts when possible.

**Comment by apoelstra** [comment]
>libcpocalypse happened precisely because libc updated something that was rarely used. If people need to upgrade fundamental crates to incompatible versions we have libcpocalypse.

They don't. Applications that want to support the new network need to switch from `Network` to `Network2`. To the extent `Network2` supports all the `Into` impls that the old one did, everything will just work. And if `Network2` *doesn't* support these, that reflects an actual breaking change associated with the new network, and the application will need to switch to a new API (which plausibly could mean switching to another ecosystem-wide fork but I'm doubtful anything would be that extreme; it seems we can always add new APIs in parallel).

Libraries meanwhile can continue to use whatever combination of `Into` impls and concrete types reflects the networks they support. And if they use the `Into` impls the type system will ensure they're not lying, unlike wildcard matches which are a runtime promise that "whatever this network is, it'll work with my library".

**Comment by Kixunil** [comment]
> We've now had several cases of downstream users complaining that `#[non_exhaustive]` broke their intended use of the `Network` enum in several ways.

Please forward those complaints to us. Just knowing the number of them and not their kind or "quality" isn't very helpful.

> Its important that there exist _some_ enum somewhere that we can use across BDK/LDK/fedi/rust-bitcoin/etc.

That can only happen if the LN developers give up their authority to define HRP for invoices to the rust-bitcoin developers. Let's say they are lazy/forgetful just like whoever didn't assign a special signet HRP for bech32, what will you do then? There logically cannot be one enum. At best there can be two: one for LN - `Currency`, one for `bitcoin`.

> If we want that to be in a separate crate that rust-bitcoin depends on with some `Into`, fine

OK, let's try to progress on this then. I'll try to split it up after `primitives` are separated if nobody beats me to it.

> there's a whole ecosystem out there, and we need to consider how best to integrate across them

Well, the funny part is if they all have the same enum they just trade immediate benefit for significant breakages later. And it's likely they don't realize it. It's the same old story about Rust: some API looks too complex and later you realize it's because of some reason you didn't even know about and it actually makes sense. And then people say things like Rust is difficult or complain at various forums where more experienced people have to explain to them that if they don't want the safety/security benefits provided by Rust maybe they should use something else... (Not trying to shame anyone here, just try to look at things from Rust perspective. I understand it takes time.)

**Comment by apoelstra** [comment]
> Please forward those complaints to us. Just knowing the number of them and not their kind or "quality" isn't very helpful.

I'm really struggling to understand this point of view and I've read this entire thread. If the enum is `#[non_exhaustive]` then it's impossible to match on it and anybody who needs to match on it is forced to go make their own enum. (Or they can stick untestable paths with unknowable semantics into their code which will become active in the future without warning.)

It seems like if we want to imagine that new future networks will match existing networks in significant ways (e.g. using bech32m but with different HRPs, using the same network protocol but with different constants) then we can make our imagining explicit by using `Into<Hrp>` and the like. In this way libraries can be generic and future-proof without eliminating a "standard set of networks enum" that can be shared amongst multiple applications and their associated libraries.

**Comment by Kixunil** [comment]
> I believe they're all in this thread, both the other issue and the several instances of LDK preferring to use `Network` (language bindings needs it, but also we'd like to drop `Currency` in `lightning-invoice`).

I mean, I would like to see specific code and more examples of it. What I want to know is if all of them are "I can't be bothered to write my own" or if there are more serious issues. It's the exact thing you pointed out: I see mainly rust-bitcoin and not that much of wider ecosystem.

> and almost certainly any one(s) that `rust-bitcoin` supports, `lightning-invoice` will too.

In real world we have a ton of incoherent garbage because someone didn't specify legacy address prefix when coming up with regtest, then someone didn't specify HRP when coming up with signet, then I remember hearing the term simnet which isn't even in Core and now I'm supposed to believe you that lightning developers are the bright exception that will swiftly standardize a unique HRP for any new network that is added to Core? Even though there are multiple teams that already have hard time agreeing on things?

I'd love to see some track record that would prove otherwise and I'd absolutely love not having 3+ different enums in the codebase but if this is the reality, rather than hiding from it and hitting problems down the line I want it represented in the type system today so I can be confident it won't break horribly in the future.

> No, they don't, this assumes some future where rust-bitcoin starts supporting a new network and all the other crates upgrade,

It doesn't need to assume the latter because if the enum is exhaustive then a **single** crate upgrading will break everything unless every single upstream crate wants to backport everything just to keep the old `Network` alive. I've seen an old codebase where lot of backporting and otherwise duplicating things led to crazy inconsistencies. Absolutely wouldn't recommend.

And even if we did this, the history says that most likely there will be some ridiculous edge case like unspecified HRP and people will start using an existing one and then things stop roundtripping and you'll be pulling your hair out trying to clean up the mess.

> I'm really struggling to understand this point of view

I know the theoretical problems with `#[non_exhaustive]` but I've seen zero lines of code where this is a practical problem. I **want** to see it, so I ask for examples. I do want to help but not over-promise in API just because someone is too lazy to write a few lines to define an enum.

> unknowable semantics

Seems pretty clear to me: only changes to how blocks headers are validated and only soft-forks when it comes to other changes. (Note that every app already must assume soft forks.) But surely, we should document that. Also even if it's untestable, anything that's not mainnet is not a big deal and mainnet is already defined.

> then we can make our imagining explicit by using `Into<Hrp>` and the like

Yes, I want that.

> libraries can be generic and future-proof

Yes, unfortunately, nothing encourages them to be - we wouldn't have this discussion otherwise. But whatever, that's on them. If `bitcoin` is solid, then I'll just use the good stuff for my projects.

**Comment by apoelstra** [comment]
>I mean, I would like to see specific code and more examples of it. 

From some test code within Liquid, setting up a bitcoin.conf file section for a specific network
```rust
        match self.network {
            Some(bitcoin::Network::Bitcoin) | None => {}, 
            Some(bitcoin::Network::Testnet) => {
                writeln!(w, "testnet=1")?;
                if version > 17_00_00 {
                    writeln!(w, "[testnet]")?;
                }   
            }   
            Some(bitcoin::Network::Signet) => {
                writeln!(w, "signet=1")?;
                if version > 17_00_00 {
                    writeln!(w, "[signet]")?;
                }   
            }   
            Some(bitcoin::Network::Regtest) => {
                writeln!(w, "regtest=1")?;
                if version > 17_00_00 {
                    writeln!(w, "[regtest]")?;
                }   
            }   
        }   
```
What should we do here in case of "unknown network"? (This is test-harness code so on some level it doesn't matter, but it's easy to imagine something like this being used in production.)

>Seems pretty clear to me: only changes to how blocks headers are validated and only soft-forks when it comes to other changes. 

So, we'd never have a test network with confidential transactions, say, or which replaced all the crypto with post-quantum equivalents, or more generally which tried some big change without bothering to define and test a soft-fork transition mechanism?

You may be correct. But it's not a promise we can make from the rust-bitcoin API.

**Comment by Kixunil** [comment]
Why not use `to_core_arg` to produce the correct string?

**Comment by apoelstra** [comment]
In this case that would happen to work, because `Bitcoin` is the only one that's a special case and it already exists. But the fact that there *is* a special case suggests that we'd rather make this code explicit rather than coercing the output of a related function.

And any future network would still need to be slightly different, because at the very least there would be no pre-0.17 syntax to generate and we'd presumably need to panic or something there.

So basically, yes, we could, but we'd still need to make assumptions about the future network, and the comments describing these assumptions would be longer and less reliable than the existing code.

**Comment by Kixunil** [comment]
If a new network is added you'll have to "panic or something" on older versions anyway so you might a s well write the code now. But I don't see a versions check for the network in the code you posted so either it's a bug in your code or you intentionally don't care about it. But also, this just gave me an idea that we should have a method returning the version Core added the network in. (Regardless of your use case.)

If you think making these assumptions is unacceptable for you then you really should make your own enum for networks which will give you option to upgrade whenever you need it rather than being forced to either change the code or miss new features in `bitcoin`. Though now I realize a common versioned `Network` is great because people who share the version don't have to write effectively the same enum multiple times.

**Comment by apoelstra** [comment]
>But I don't see a versions check for the network in the code you posted so either it's a bug in your code or you intentionally don't care about it.

A version check for what network?

> Though now I realize a common versioned `Network` is great because people who share the version don't have to write effectively the same enum multiple times.

Yes. If we were using `Into<Hrp>` etc in this library then the `Network` enum literally wouldn't matter to rust-bitcoin because we'd never be directly using it here. Its only purpose is to provide a canonical "these are the networks that the rust-bitcoin ecosystem supports" indicator and to provide a schelling point for other developers to use as a set of supported networks.

This is why I continue not to see any "libcopalypse" scenario happening, nor do I see a problem with just adding a `Network2` enum if we want to widen the set of networks.

**Comment by Kixunil** [comment]
> A version check for what network?

Something like `fn is_supported_by_core_version(self, core_version: u32) -> bool`

> This is why I continue not to see any "libcopalypse" scenario happening,

It's less of a problem if we do it right and also less of a problem for consumers of downstream crates that do it right. I'm not hyped enough to say it's non-issue in that case but it might be.

**Comment by apoelstra** [comment]
Core is never going to update so that existing configuration files stop working. So I'm not sure what a Core version check would accomplish.

>It's less of a problem if we do it right

If "do it right" means putting `#[non_exhaustive]` on it, then nobody will use the `Network` type included in rust-bitcoin so we might as well just drop it.

**Comment by Kixunil** [comment]
> So I'm not sure what a Core version check would accomplish.

Very similar thing your `version > 17_00_00` check does: raise a readable error if the core version doesn't support specific network. (In your case you don't need to error.)

> If "do it right" means putting `#[non_exhaustive]` on it,

No, it's doing all these traits and making sure nothing returns it etc...

**Comment by apoelstra** [comment]
>No, it's doing all these traits and making sure nothing returns it etc...

Ok, perfect, we are in agreement there :).

So if we do things right in rust-bitcoin, the `Network` enum actually will be unsused within rust-bitcoin, other than abstractly as a "Into<Params>" or "Into<Hrp>" or whatever. This gives us a lot of freedom here. We can

* Have no blessed `Network` enum at all (personally I could live with this; for my own projects I never need the enum to be interoperable; it sounds like the ldk people would need to define their own enum but they already have a near-universal `rust-lightning` dep so they could put it there).
* Have a `#[non_exhaustive]` `Network` enum whose documentation explains how to match on it (and we will have a drawn-out fight over what this text should be).
* Have an exhaustive enum which we will version by releasing rust-bitcoin 2.0 whenever a new network arrives. (I think we all agree this is a bad idea; it would definitely cause libcopalypse).
* Have an exhaustive enum in its own crate, which gets upgraded by major versions. We advise all *libraries* to never return the type, accept it only via generics, to access the type through rust-bitcoin, and failing that, accept all published major versions. rust-bitcoin naturally would do all of these things.
* Have an exhaustive enum which we will version by just adding a parallel `Network2 enum. Again, tell *libraries* to avoid directly using the type if at all possible.

In fact, we could do any combination of these since none of them actually interfere with each other (though it would probably be confusing!).

**My personal view** at this point is
* if we're telling libraries to basically not use the enum, then exhaustiveness becomes irrelevant for libraries (so no risk of libcopalypse) but remains relevant for applications (who want exhaustiveness so that their users aren't suddenly using a new network that the application authors don't know about and can't predict behavior of)
* having said that , there are some "applications" (e.g. the Lightning Network) which are spread across multiple libraries and actual applications, and they actually *will* use the enum despite our advice (and I get why that makes you nervous @Kixunil since then our type is implicitly reflecting some LN thing that we have no relationship to). I think this is still fine; if we later add an enum that they don't like, or upgrade our type in a breaking way, they can have a "local" libcopalypse but it won't affect people outside their space

I continue to think we should have an exhaustive enum, with a bunch of documentation about how to use it properly. I am warming up to the "don't have an enum, just have rules for defining your own" but I think this would be needless work for people who just want to grab an off-the-shelf set of networks or people who want to make their own enum but want to see an example of how to do it first.

**Comment by Kixunil** [comment]
Yeah, pretty much it. I want combination of having `#[non_exhaustive]` enum in some reasonable crate and an exhaustive one in a separate versioned crate. With all the docs you mention and the enums mentioning each-other.

> I think this would be needless work for people who just want to grab an off-the-shelf set of networks or people who want to make their own enum but want to see an example of how to do it first.

Yep.

**Comment by tcharding** [comment]
Cool, if I'm following, we have rough consensus here now, and a solution on how to progress. We can get this done before next release, right?

@TheBlueMatt  are you in agreement here or have we missed something? If we haven't I'll get to work.

That would still make v31 annoying for users though.

**Comment by Kixunil** [comment]
> We can get this done before next release, right?

I doubt so, I suspect we need to progress with crate splitting more. But if you find a way to do it I'm fine with it.

> Can we resolve that, remove `non_exhaustive` in a point release?

And we'll be putting it back in the next one and then people will have to fix their code again. It's not really worth it.

**Comment by tcharding** [comment]
Fair and fair. I'll just do nothing then :)

**Comment by apoelstra** [comment]
What would the crate dep tree look like here? Would we have two separate crates and rust-bitcoin depends on both of them? Where does `Params` live and where does the `Into<Params>` impl live?

**Comment by Kixunil** [comment]
Currently I'm not sure if the `non_exhaustive` enum should go to a separate crate or it fits somewhere else, I'd like to first split up other things. I also suspect `bitcoin` can't depend on exhaustive one but the exhaustive one can depend on `bitcoin`. `Params` will probably live alongside with the `non_exhaustive` one.

I think we'll see it more clearly when we're closer. There's other stuff to do now.

**Comment by apoelstra** [comment]
Ok, so crate(s) that depend on rust-bitcoin, rather than the other way around. Ok, makes sense.

I suspect both the non-exhaustive and the exhaustive one could go into the same crate -- aside from naming conflicts. Because the non_exhaustive type can be expanded in minor versions, and when the exhaustive type forces major versions, we could semver-trick the non-exhaustive one to prevent breakage. Maaaybe we could even use the same type but have exhaustiveness be feature-gated (this would be a super weird feature-gate that'd turn off the semver trick, so maybe it just doesn't work).

Anyway agreed that there are higher priorities right now and happy to revisit when we're closer.

**Comment by Kixunil** [comment]
> both the non-exhaustive and the exhaustive one could go into the same crate

They can't because they have to be versioned differently. Semver trick would have to maintain old versions indefinitely and cause crate duplication which would trigger CI checks and people would then have to add exceptions. Semver trick is great but not having to use it in the first place is better.

> exhaustiveness be feature-gated

Strictly speaking, removing `non_exhaustive` is not a breaking change but triggers a bunch of warnings about unreachable branches so I'd rather avoid that.

**Comment by tcharding** [comment]
Removing the milestone from this issue, we have #2541 to hopefully alleviate the current pain.

**Comment by Kixunil** [comment]
Given a move to solve this was requested, here's my proposal:

* Create a crate called `bitcoin-params` that contains the `non_exhaustive` `Network` enum with comprehensive documentation about how to deal with it (see below), as well as `Params` and possibly other things.
* Create a crate called `bitcoincore-network` that contains an exhaustive `Network` enum the current set of networks supported by Bitcoin Core
* Make sure all relevant conversions are implemented (in whichever crate it makes sense)

I'm not really sure how to handle address though. It will work fine for bech32 but if anything new is ever created I have no idea how will we sync them up. I will definitely think about this.

Here's proposed documentation of the enum:

## Dealing with Bitcoin Network

When an application wants to support multiple networks (as it should to allow testing), there are generally two possible approaches:

* Make the entire code network-agnostic: don't manually match on the `Network` enum and only use the upstream methods to serialize/deserialize it (as string or using magic in binary)
* Decide on some specific set of networks that the application supports and don't support anything else.

Ideally, the first approach would be used by everyone because it allows trivial upgrades but that is not always possible. When dealing with existing protocols that define their own ways of encoding the network a new network cannot be added until the standard is updated. E.g. such is the case of Lightning Network which defined its own serialization of the network as human-readable string and the standard doesn't say anything about future networks.

However each application already has to deal with some kind of "unsupported network" handling. This is done when parsing the inputs to the application - e.g. if an application receives an argument "satoshi" which is not a valid name of the network this has to be rejected. Similarly if some binary identifier (like magic) is used an invalid value needs to returns error. And this is also the natural place to check that the network is known but not supported by an application. You can then carry a proof that you did this check throughout the entire application by converting the network into a custom enum that only contains variants for networks your application supports. And because the popular sets of supported networks are "whatever Bitcoin Core supports" we made the `bitcoincore-network` crate storing these, so you don't have to write the enum yourself if you are in that situation.

Note though that using an existing enum in public interfaces has the specific downside that you won't be able to change the set of supported networks without breaking the API. For instance, if a lightning library uses one of these pre-made enums and a new network is both released by Core and described in LN specification then the said lightning library cannot add it without breaking its users. However it can still defend against this by making its own enum `#[non_exhaustive]`! Thus this is the recommended approach for libraries. (They can still depend on the crate to provide conversions from the exhaustive enums.)

## Upgrading from previous versions of `bitcoin`

*If you've skipped the above section, please read it first.*

In the past the `Network` enum was erroneously exhaustive. This made some applications accidentally write a code as if it was assuming the `bitcoin` crate won't ever support new networks. That means maybe they have code that matches on the network exhaustively. In some scenarios it might be possible to remove it (update the protocol to support future networks) and that is the preferred approach. However if it's not possible the only solution is to use the approach above - use a pre-defined or custom enum specifying the exact supported networks.

You can do this by replacing the existing matches to take your type instead and then just following the compiler errors to tell you which types need to be changed. When you reach the place where you're parsing/decoding the network you instead write a conversion from this crate's enum to yours returning an error if the network is unsupported - just as you're already returning an error when you receive garbage input.

### Example

If you have this code:

```rust
use bitcoin::Network;

fn foo(network: Network) {
    match network {
        Network::Bitcoin => ...,
        Network::Testnet => ...,
        Network::Regtest => ...,
        Network:Signet => ...,
    }
}

mod internal {
    use bitcoin::Network;
    fn bar(network: Network) { super::foo(network) }
}

mod parse {
    use bitcoin::Network;

    fn baz(arg: &str) -> anyhow::Result<()> {
        super::internal::bar(arg.parse::<Network>()?)
        Ok(())
    }
}
```

You can change it to:

```rust
// we've changed use bitcoin::Network here to this
// note that we've named it Network to minimize the changes
#[non_exhaustive]
enum Network {
    Bitcoin,
    Testnet,
    Regtest,
    Signet,
}

impl TryFrom<bitcoin::Network> for Network {
    type Error = anyhow::Error;

    fn try_from(value: bitcoin::Network) -> Result<Self, Self::Error> {
        match value {
            bitcoin::Network::Bitcoin => Ok(Network::Bitcoin),
            bitcoin::Network::Testnet => Ok(Network::Testnet),
            bitcoin::Network::Regtest => Ok(Network::Regtest),
            bitcoin::Network::Signet => Ok(Network::Signet),
            unsupported => Err(anyhow::anyhow!("Unsupported bitcoin network: {}", unsupported)),
        }
    }
}

// continues working!
fn foo(network: Network) {
    match network {
        Network::Bitcoin => ...,
        Network::Testnet => ...,
        Network::Regtest => ...,
        Network:Signet => ...,
    }
}

mod internal {
    // Here we've only needed to change the import to use our enum instead of bitcoin enum
    use super::Network;
    fn bar(network: Network) { super::foo(network) }
}

mod parse {
    use bitcoin::Network;

    fn baz(arg: &str) -> anyhow::Result<()> {
        // We've added conversion here
        super::internal::bar(arg.parse::<Network>()?.try_into()?)
        Ok(())
    }
}
```

Similarly, if you're e.g. deserializing a custom `u8`, you do the above but instead of conversion you deserialize directly into your own enum.

**Comment by apoelstra** [comment]
> * Create a crate called `bitcoin-params` that contains the `non_exhaustive` `Network` enum with comprehensive documentation about how to deal with it (see below), as well as `Params` and possibly other things.

Why should `Params` be moved out of `bitcoin`? It is more important that `Params` stay stable than that `Network`  does. Also, would `bitocin-params` need to depend on `addresses` and other crates so that it can impl `Into<NetworkKind>` for `Params` and its network enum?
  
> * Create a crate called `bitcoincore-network` that contains an exhaustive `Network` enum the current set of networks supported by Bitcoin Core

Why not just use `bitcoin` for this? For the forseeable future we will release new `bitcoin` versions far more frequently that new networks will appear, and with `units` and `primitives` and `addresses` split off there will be minimal breakage for users who want to keep up with it.

But if we did split off this crate, would *it* need to depend on `addresses`?

**Comment by Kixunil** [comment]
> Why should `Params` be moved out of `bitcoin`?

I think it makes a lot of sense to have `Params` and `Netwrok` together since `Network` is convertible to `Params` and technically `Params` could have `network_category(&self) -> Network` (where we ignore `Signet` challenge script). That creates a dependency loop.

> It is more important that `Params` stay stable than that `Network` does.

I guess, but that actually makes even stronger case for moving `Params` out! Note that `Address` needs to take *something* to define the address prefixes and whatnot.

> `bitocin-params` need to depend on `addresses`

No, it's the other way around

> Why not just use `bitcoin` for this?

Heh, that could work indeed but there are a few things I don't like about it:

* It bundles various breaking changes together for different reasons.
* If we know already that once `bitcoin` reaches 1.0 the enum will *have to* move to not destabilize the crate then doing it later, when more code depends on it, results in more overall work than doing it sooner
* Code being in root of `bitcoin` may be seen as "promoted" and using that enum is not really what the default should be. The real default is the `non_exhaustive` enum and anything else is fallback for cases when it's not possible (existing protocols that use their own constants for different networks).

> But if we did split off this crate, would _it_ need to depend on `addresses`?

No, but it should (maybe optionally; but I think rather not) depend on the `non_exhaustive` enum so that conversions can be implemented and all getters can be also implemented by converting to the `non_exhaustive` first and then calling the respective getter. This would ensure consistency and maybe it can be done with just a macro.

**Comment by tcharding** [comment]
Closed by #4640
